### PR TITLE
Fixes #17521 - Use short name for hosts

### PR DESCRIPTION
**Pull request by parthaa**
This commit enforces the policy of using short name for hosts
while preserving the fqdn for interface values provided the user has
enabled 'use_shortname_for_vms' Setting.

### CODE REVIEW

**Comment by ohadlevy** [comment]
@parthaa mind rebasing your PR? test failure seems related to a previous fix.

**Comment by ares** [comment]
should also `sync_required?` be modified to reflect this? also please add tests to https://github.com/theforeman/foreman/blob/develop/test/unit/name_synchronizer_test.rb

**Comment by parthaa** [comment]
> should also sync_required? be modified to reflect this? also please add tests to

updated. added tests too..

**Comment by parthaa** [comment]
Based on comments by @domcleal at http://projects.theforeman.org/issues/17521#note-2
I created a setting "append_domain_name_to_hosts"  to make this work instead of use_shortname_for_vms . I have updated the tests accordingly. 

@ohadlevy @ares @domcleal  @dLobatog really need this to be reviewed asap.. High priority on downstream.

File: app/models/lookup_value.rb
Comment (ares) [inline app/models/lookup_value.rb]: If users prefer to use short names, shouldn't the matchers also use short names?

If we keep this change, I think the order should be the other way around, first we should search the host by fqdn and then fallback to facts. I see two reasons for this, first that the host name would have higher priority in case these two values get out of sync so it would prefer current behavior, second is that the search for is more resource consuming while the host name will be mostly used.

Also shouldn't the search_for be called on `Host.unscoped`?

Comment (ares) [reply app/models/lookup_value.rb]: This change also needs to be covered by tests

File: app/services/name_synchronizer.rb
Comment (ares) [inline app/services/name_synchronizer.rb]: please extract this to some private method so it's not duplicated in `sync_name` and `sync_required?`

**Review by ares — CHANGES_REQUESTED** [review]
Could you please share more information why this is needed? I don't like the fact that we'd store different value in host name based on settings. What if user later changes the setting value? There's no simple way to update all hosts right?

**Comment by parthaa** [comment]
> Could you please share more information why this is needed? I don't like the fact that we'd store different value in host name based on settings. What if user later changes the setting value? There's no simple way to update all hosts right?

Entire point of this PR is because some customers do not want domain names appended to the host name *even* if they choose an interface. They *always* want the short name. Since thats the not the default setting for foreman we need this setting.

**Comment by parthaa** [comment]
> If we keep this change, I think the order should be the other way around, first we should search the host by fqdn and then fallback to facts. I see two reasons for this, first that the host name would have higher priority in case these two values get out of sync so it would prefer current behavior, second is that the search for is more resource consuming while the host name will be mostly used.

Lets start with
https://github.com/theforeman/foreman/blob/develop/app/models/host/base.rb#L306

So when you do a Host.save with a overridden smart parameter, the host creates a "LookupValueMatcher" that has a match attribute  ```fqdn=<fqdn> || name ```  

When the newly created LookupValue tries to get saved it hits -> https://github.com/theforeman/foreman/blob/develop/app/models/lookup_value.rb#L73
" Host.unscoped.find_by_name(fqdn) " 

This causes the lookup fail and the host save to thus fail.. Basically we need a way to change this code to say the equivalent of
"Host.unscoped.find_by_fqdn(fqdn)" instead of " Host.unscoped.find_by_name(fqdn) " 
Changes in that file are meant to address that.

Order wise I think we have it better here. Because if the LookupValueMatcher has ```fqdn=<fqdn> || name ```  one would think fqdn gets the higher priority.  I can see the "more resources" argument though. I changed it anyway.

**Comment by parthaa** [comment]
> Also shouldn't the search_for be called on Host.unscoped?

changed. Good call

File: app/services/name_synchronizer.rb
Comment (ares) [inline app/services/name_synchronizer.rb]: I'd avoid creating a variable here, how about this

```ruby
@host.send :write_attribute, :name, interface_name
```

File: app/services/name_synchronizer.rb
Comment (ares) [inline app/services/name_synchronizer.rb]:
```ruby
 @interface.primary? && @host.present? && (@host.name != interface_name)
```

File: app/models/setting/provisioning.rb
Comment (ares) [inline app/models/setting/provisioning.rb]: I don't think this is specific to provisioning. The same would apply when new hosts registers by facts upload or is created by API call. I think it should go to general settings and the help should be reworded to something like "Append domain names to new hosts".

**Review by ares — DISMISSED** [review]
Sorry, it does not work. When I disable domain appending and I override some smart class parameter, I see the default value in host edit form in Parameters tab correctly. But when I define matcher in smart class parameter form with this fqdn equal to this short name without domain, the new value is not displayed in the Parameters tab in the host form. I still see only the default one. Also when I try to override it from host form, once I click submit I'm redirected back to host form without displaying any error message, in logs I can find following

```
2017-01-18T16:52:39 5b2b4524 [app] [D] Ignoring #<LookupValue id: nil, match: "fqdn=joy-ruffins.example.tst", value: "bbb", lookup_key_id: 157, created_at: nil, updated_at: nil, omit: false> match errors: ["fqdn=joy-ruffins.example.tst does not match an existing host"]
```

it tries to create the matcher with fqdn instead of hostname.

Also when I switch the setting to apply domains again, it leaves all matchers with short names making them invalid when I try to save the smart class parameter. We probably have the same problem with other renaming so maybe that is acceptable. Also the host correctly changes name after settings switch and new save.

File: app/models/setting/provisioning.rb
Comment (ares) [reply app/models/setting/provisioning.rb]: Since it applies whenever the host is saved maybe the wording should not mention new hosts but something like "Append domain names to host names upon save"

Comment (dLobatog) [reply app/models/setting/provisioning.rb]: 👍 "Append domain names to new hosts" - then explain what it means on the description sounds good

Comment (ares) [reply app/models/setting/provisioning.rb]: it's not only for new hosts (my second comment), it applies whenever the host is saved and if the setting changed, save can append or remove the domain e.g. when I change the owner

**Comment by parthaa** [comment]
> Sorry, it does not work. When I disable domain appending and I override some smart class parameter, I see the default value in host edit form in Parameters tab correctly. But when I define matcher in smart class parameter form with this fqdn equal to this short name without domain, the new value is not displayed in the Parameters tab in the host form. I still see only the default one. Also when I try to override it from host form, once I click submit I'm redirected back to host form without displaying any error message, in logs I can find following

@ares @dLobatog 

So I am guessing might have happened is you the override set to  match based on shortname instead of using fqdn even as the matcher said it was going to use fqdn. Would you mind confirming this again. When setting up the override matcher please set the full hostname instead of just the shortname.

**Comment by ohadlevy** [comment]
@ares what is the status of this PR ? thanks?

File: app/models/setting/provisioning.rb
Comment (dLobatog) [reply app/models/setting/provisioning.rb]: Correct, it should probably be moved to general.

File: app/models/lookup_value.rb
Comment (dLobatog) [inline app/models/lookup_value.rb]: This is not really user-friendly, why not adding a new matcher 'shortname'? Maybe it'd be easy to add this matcher in another PR then use it here, just to not mix stuff and keep PRs simple.

**Review by dLobatog — CHANGES_REQUESTED** [review]
@parthaa 👍 to the setting, and the name synchronizer seems to work pretty well for created and imported hosts.  

The fqdn matcher though, it's counterintuitive IMO. I would suggest leaving fqdn matcher as it stands and making a new shortname one.. not sure about other people's thoughts on it here. 

The new `shortname` matcher would be useful for people that mix shortnames/fqdns too, and it could be documented on it's own without risking confusing people. Also it'd allow to merge part of this PR right away.

Thanks @parthaa !

**Comment by iNecas** [comment]
@parthaa what is the status of this PR?

File: app/models/lookup_value.rb
Comment (parthaa) [reply app/models/lookup_value.rb]: > This is not really user-friendly, why not adding a new matcher 'shortname'? Maybe it'd be easy to add this matcher in another PR then use it here, just to not mix stuff and keep PRs simple.


@dlobatog Ok let me see if I can state this in a clearer fashion. That line in lookup value is deceptively simple looking.

The bug is that Customer wants short hostnames when using the API But wants to have a fully formed fqdn in their puppet fact. They were able to get away with this setup before Host unification with Katello because Content Host and Host were 2 different things. However when they unified they wanted the host to have the short host name. But they also wanted the fqdn fact in puppet to reflect fully qualified name including the interface domain name.

Not just that. On top of this they want to be able Override certain puppet configuration (aka smart class parameters) based on the FQDN
Foreman code prior to this change made the assumption that Host.Name == FQDN when matching.
That is why we see code like
```return true if Host.unscoped.find_by_name(fqdn) || host_or_hostgroup.try(:new_record?) ||```

This works fine in most cases because foreman always appends interface name. However it would be better if in addition to matching host names for fqdn it also  tries to identify/find/match hosts with their actual fqdn value as specified in their puppet fact. 

All my code does -> When Matching..
- Try matching by the existing method first. i.e See if you can find a host that has name given in the FQDN matcher . Identical to the line above
-  If that does not exist See if there is a host that has the FQDN puppet fact matching the value given in the FQDN matcher
that 2nd case is where this customer's matcher would match.

I would argue that this change improves the existing 'matching' setup by actually matching fqdns for real instead of assuming name given the foreman host == FQDN

Adding a shortname matcher is not really going to solve the issue for this customer because they want the existing FQDN matcher to actually match FQDN. Not name.

**Comment by parthaa** [comment]
> Correct, it should probably be moved to general.
Done

File: app/models/lookup_value.rb
Comment (ares) [reply app/models/lookup_value.rb]: If customer needs to combina short name and FQDN in single matcher, I think a new matcher should be added. Something like `fqdn_or_shortname`. We're trying to improve project usability, this would go the opposite direction. I think the best solution is to add `shortname` and `fqdn_or_shortname` to already existing `fqdn` and leave `fqdn` untouched.

**Comment by parthaa** [comment]
@dLobatog @ares created -> http://projects.theforeman.org/issues/21929 and updated this.
Thanks for reading my comments earlier.

**Review by dLobatog — APPROVED** [review]
@parthaa 👍 from my side to the current changes, I think it's clear that if you change the setting, that only applies to new hosts. 

@ares any further comments?

**Comment by ares** [comment]
I did not test it, I still think this is a bit confusing especially if the setting is global it does not address the use case where some hosts have FQDN and some not. But I'm happy to dismiss my review so it's not blocked on me finding time to do another round of review.

**Review by dLobatog — APPROVED** [review]
Thanks @parthaa  - let's follow up this with http://projects.theforeman.org/issues/21929 when you or someone else gets a chance.

**Comment by ares** [comment]
Thanks, setting the release to 1.18 in redmine, please don't forget to update redmine issue after merging.

**Pull request merged**
